etl/analytics.py

```diff
@@ -0,0 +1,833 @@
+    df_views["n_days"] = (pd.to_datetime(date_max) - start_date).dt.days
+
+    # Add a column for the average number of daily views.
+    df_views["views_daily"] = df_views["views"] / df_views["n_days"]
+
+    # Fix infs (for charts that were published in the last day).
+    df_views.loc[df_views["views_daily"] == float("inf"), "views_daily"] = 0
+
+    return df_views
+
+
+def get_chart_views_last_n_days(
+    chart_ids: Optional[List[int]] = None,
+    n_days: int = 30,
+) -> pd.DataFrame:
+    """
+    Fetch number of chart views per chart for the last n_days.
+
+    Parameters
+    ----------
+    chart_ids : list of int, optional
+        List of chart IDs to filter the results. If None, all charts are included.
+    n_days : int
+        Number of days to look back for views. Default is 30 days.
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame containing the number of chart views per chart for the last n_days.
+
+    """
+    # Calculate date range.
+    date_max = str(datetime.today().date())
+    date_min = str((datetime.today() - pd.Timedelta(days=n_days)).date())
+
+    # Get views.
+    df_views = get_chart_views_by_chart_id(chart_ids=chart_ids, date_min=date_min, date_max=date_max)
+
+    return df_views
+
+
+def get_post_views_by_url(
+    urls: Optional[List[str]] = None,
+    date_min: str = DATE_MIN,
+    date_max: str = DATE_MAX,
+) -> pd.DataFrame:
+    """
+    Fetch number of GDoc views per URL from Metabase.
+
+    Parameters
+    ----------
+    urls : list of str, optional
+        List of URLs to filter the results. If None, all URLs are included.
+    date_min : str, optional
+        Minimum date to filter the results. If None, no minimum date is applied.
+    date_max : str, optional
+        Maximum date to filter the results. If None, no maximum date is applied.
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame containing the number of GDoc views per URL.
+
+    """
+    where_clauses = []
+    if date_min:
+        where_clauses.append(f"day >= '{date_min}'")
+    if date_max:
+        where_clauses.append(f"day <= '{date_max}'")
+    if urls:
+        url_list = ", ".join(f"'{url}'" for url in urls)
+        where_clauses.append(f"url IN ({url_list})")
+    # Exclude pages corresponding to grapher charts and explorers.
+    where_clauses.append("url NOT LIKE '%/grapher/%'")
+    where_clauses.append("url NOT LIKE '%/explorers/%'")
+    # Exclude url with spaces or empty.
+    where_clauses.append("url NOT LIKE '% %'")
+    where_clauses.append("url IS NOT NULL")
+    where_sql = "WHERE " + " AND ".join(where_clauses) if where_clauses else ""
+
+    query = f"""
+    SELECT
+        url,
+        SUM(views) AS views
+    FROM views_detailed
+    {where_sql}
+    GROUP BY url
+    ORDER BY views DESC
+    """
+    df_views = read_metabase(sql=query)
+
+    n_days = (pd.to_datetime(date_max) - pd.to_datetime(date_min)).days
```

Comment:
same issue as previous comments that `date_max` might be later than the max date of data available.